### ISSUE TITLE
GlobalOpentelemetry.set has already been called

### ISSUE THREAD

**Issue by roychshao** [open-telemetry/opentelemetry-java#5343]
hello!
I am trying to add opentelemetry into a maven project, I want to manually use it and I have some problems.
At first, I follow the document: https://opentelemetry.io/docs/instrumentation/java/manual/ to add the dependencies on my project.
And then config like the document.(I didn't import some about metrics since I only need trace)
below is how I config it.
```java
package org.mybatis.jpetstore.domain;

import io.opentelemetry.api.OpenTelemetry;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator;
import io.opentelemetry.context.propagation.ContextPropagators;
import io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporter;
import io.opentelemetry.sdk.OpenTelemetrySdk;
import io.opentelemetry.sdk.resources.Resource;
import io.opentelemetry.sdk.trace.SdkTracerProvider;
import io.opentelemetry.sdk.trace.export.BatchSpanProcessor;
import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;

public class Tracing {
  public Resource resource;
  public SdkTracerProvider sdkTracerProvider;
  public OpenTelemetry openTelemetry;
  private Tracer tracer;

  public Tracing() {
    // Create a default resource with a service name attribute
    Resource defaultResource = Resource.getDefault();
    Resource serviceNameResource = Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME, "jpetstore-main"));
    this.resource = defaultResource.merge(serviceNameResource);

    // Create a tracer provider with a batch span processor and the given resource
    BatchSpanProcessor spanProcessor = BatchSpanProcessor.builder(OtlpGrpcSpanExporter.builder().build()).build();
    this.sdkTracerProvider = SdkTracerProvider.builder().addSpanProcessor(spanProcessor).setResource(resource).build();

    // Create an OpenTelemetry instance with the given tracer provider and propagator
    W3CTraceContextPropagator propagator = W3CTraceContextPropagator.getInstance();
    ContextPropagators propagators = ContextPropagators.create(propagator);

    this.openTelemetry = OpenTelemetrySdk.builder().setTracerProvider(sdkTracerProvider).setPropagators(propagators)
        .buildAndRegisterGlobal();

    this.tracer = this.openTelemetry.getTracer("jpetstore-main", "1.0.0");
    System.out.println(tracer);
  }

  public Tracer getTracer() {
    return this.tracer;
  }
}

```
I create a class named Tracing and I want to config all about opentelemetry here, so that I can just import this and use it every other place.
however, the errors happens.
```
Caused by: java.lang.IllegalStateException: GlobalOpenTelemetry.set has already been called. GlobalOpenTelemetry.set must be called only once before any calls to GlobalOpenTelemetry.get. If you are using
 the OpenTelemetrySdk, use OpenTelemetrySdkBuilder.buildAndRegisterGlobal instead. Previous invocation set to cause of this exception.                                                                     
        at io.opentelemetry.api.GlobalOpenTelemetry.set(GlobalOpenTelemetry.java:103)                                                                                                                      
        at io.opentelemetry.sdk.OpenTelemetrySdkBuilder.buildAndRegisterGlobal(OpenTelemetrySdkBuilder.java:85)                                                                                            
        at org.mybatis.jpetstore.domain.Tracing.<init>(Tracing.java:51)                                                                                                                                    
        at org.mybatis.jpetstore.service.CatalogService.<init>(CatalogService.java:43)                                                                                                                     
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)                                                                                                              
        at org.mockito.internal.util.reflection.InstrumentationMemberAccessor$Dispatcher$ByteBuddy$yQ4WRpst.invokeWithArguments(Unknown Source)                                                            
        at org.mockito.internal.util.reflection.InstrumentationMemberAccessor.lambda$newInstance$0(InstrumentationMemberAccessor.java:199)                                                                 
        at org.mockito.internal.util.reflection.InstrumentationMemberAccessor.newInstance(InstrumentationMemberAccessor.java:188)                                                                          
        at org.mockito.internal.util.reflection.InstrumentationMemberAccessor.newInstance(InstrumentationMemberAccessor.java:161)                                                                          
        at org.mockito.internal.util.reflection.ModuleMemberAccessor.newInstance(ModuleMemberAccessor.java:42)                                                                                             
        at org.mockito.internal.util.reflection.FieldInitializer$ParameterizedConstructorInstantiator.instantiate(FieldInitializer.java:291)                                                               
        at org.mockito.internal.util.reflection.FieldInitializer.acquireFieldInstance(FieldInitializer.java:146)                                                                                           
        at org.mockito.internal.util.reflection.FieldInitializer.initialize(FieldInitializer.java:91)                                                                                                      
        at org.mockito.internal.configuration.injection.ConstructorInjection.processInjection(ConstructorInjection.java:48)                                                                                
        at org.mockito.internal.configuration.injection.MockInjectionStrategy.process(MockInjectionStrategy.java:68)                                                                                       
        at org.mockito.internal.configuration.injection.MockInjectionStrategy.relayProcessToNextStrategy(MockInjectionStrategy.java:91)                                                                    
        at org.mockito.internal.configuration.injection.MockInjectionStrategy.process(MockInjectionStrategy.java:71)                                                                                       
        at org.mockito.internal.configuration.injection.MockInjection$OngoingMockInjection.apply(MockInjection.java:88)                                                                                    
        at org.mockito.internal.configuration.DefaultInjectionEngine.injectMocksOnFields(DefaultInjectionEngine.java:26)                                                                                   
        at org.mockito.internal.configuration.InjectingAnnotationEngine.injectCloseableMocks(InjectingAnnotationEngine.java:118)                                                                           
        at org.mockito.internal.configuration.InjectingAnnotationEngine.processInjectMocks(InjectingAnnotationEngine.java:61)                                                                              
        at org.mockito.internal.configuration.InjectingAnnotationEngine.process(InjectingAnnotationEngine.java:48)                                                                                         
        at org.mockito.MockitoAnnotations.openMocks(MockitoAnnotations.java:81)                                                                                                                            
        at org.mockito.internal.framework.DefaultMockitoSession.<init>(DefaultMockitoSession.java:43)                                                                                                      
        at org.mockito.internal.session.DefaultMockitoSessionBuilder.startMocking(DefaultMockitoSessionBuilder.java:83)                                                                                    
        ... 56 more                                                                                                                                                                                        
Caused by: java.lang.Throwable                                                                                                                                                                             
        at io.opentelemetry.api.GlobalOpenTelemetry.set(GlobalOpenTelemetry.java:111)                                                                                                                      
        ... 80 more
```
these rows are the problem itself: 
```java
this.openTelemetry = OpenTelemetrySdk.builder().setTracerProvider(sdkTracerProvider).setPropagators(propagators)
    .buildAndRegisterGlobal();
``` 
It seems like the GlobalOpenTelemetry.set has been called before, but I do not call that, I call buildAndRegisterGlobal instead.
I want to know the if I loss something or do I config it wrong?
If not, how this error happens and are there any solution for this situation?

I am a rookie and new to opentelemetry-java
hope for help!! thanks!

**Comment by trask** [comment]
hi @roychshao!

Since you are running into issues with unit tests, it may be helpful to use `GlobalOpenTelemetry.resetForTest()` in your test setup or tear down methods to clear the global OpenTelemetry instance before running each test. This will allow you to create a new OpenTelemetry instance for each test and avoid interference between tests.

You can call the `resetForTest()` method before or after each test to clear the global instance and avoid interference between tests:

```java

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

public class MyTest {
  @Before
  public void setUp() {
    GlobalOpenTelemetry.resetForTest(); // Reset the global instance before each test
  }

  @After
  public void tearDown()
```

**Comment by roychshao** [comment]
> another option is to avoid `buildAndRegisterGlobal` and just use `build`

Thanks for your help!
It really helps a lot!

**Comment by roychshao** [comment]
@trask Excuse me,
I have another question, are there any api or library in Java like AsyncHooksContextManager in Javascript?

**Comment by jkwatson** [comment]
closing this issue as resolved

**Comment by davidmichaelkarr** [comment]
I'm also seeing this problem, but adding the "resetForTest" call makes no difference.  I even tried putting it in both the the BeforeEach AND the AfterEach methods.  No difference.  However, I also note that I don't get this error when I run the test in Eclipse, but I do get the error when I run it from a mvn build.

**Comment by trask** [comment]
Hi @davidmichaelkarr, can you open a new issue and provide more details? Thanks
